File: src/main/java/me/alegian/thavma/impl/client/renderer/AspectRenderer.java
package me.alegian.thavma.impl.client.renderer;

import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import me.alegian.thavma.impl.Thavma;
import me.alegian.thavma.impl.client.T7GuiGraphics;
import me.alegian.thavma.impl.client.texture.atlas.AspectAtlas;
import me.alegian.thavma.impl.client.util.GuiGraphicsExtensionsKt;
import me.alegian.thavma.impl.common.aspect.Aspect;
import me.alegian.thavma.impl.common.aspect.AspectMap;
import me.alegian.thavma.impl.common.aspect.AspectStack;
import net.minecraft.client.Camera;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.core.BlockPos;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;
import org.joml.Vector2f;

@OnlyIn(Dist.CLIENT)
public class AspectRenderer {
  public static final int ROW_SIZE = 5;
  public static final float QUAD_SIZE = .3f; // aspect width is .3 blocks
  private static final int PIXEL_RESOLUTION = 16; // pixels per block

  /**
   * Renders the Aspect contents of an AspectContainer, using GuiGraphics.<br>
   * GuiGraphics assumes integer pixel coordinates, so we multiply everything by 16 (so a block is 16x16).
   */
  public static void renderAfterWeather(AspectMap aspects, PoseStack poseStack, Camera camera, BlockPos blockPos) {
    if (aspects.isEmpty()) return;

    poseStack.pushPose();
    var cameraPos = camera.getPosition();
    poseStack.translate(blockPos.getX() - cameraPos.x() + 0.5d, blockPos.getY() - cameraPos.y() + 1.25d + AspectRenderer.QUAD_SIZE / 2, blockPos.getZ() - cameraPos.z() + 0.5d);
    var angle = RenderHelperKt.calculatePlayerAngle(blockPos.getCenter());
    poseStack.mulPose(Axis.YP.rotation(angle));
    poseStack.scale(AspectRenderer.QUAD_SIZE, AspectRenderer.QUAD_SIZE, 1); // this puts us in "aspect space" where 1 means 1 aspect width

    T7GuiGraphics guiGraphics = new T7GuiGraphics(Minecraft.getInstance(), poseStack, Minecraft.getInstance().renderBuffers().bufferSource());

    // these offsets account for wrapping to new lines, and centering the aspects
    Vector2f[] offsets = AspectRenderer.calculateOffsets(aspects.size());

    int i = 0;
    for (AspectStack aspectStack : aspects.displayedAspects()) {
      poseStack.pushPose();
      poseStack.translate(offsets[i].x, offsets[i].y, 0);

      // gui rendering is done in pixel space
      poseStack.scale(1f / AspectRenderer.PIXEL_RESOLUTION, -1f / AspectRenderer.PIXEL_RESOLUTION, 1f);
      AspectRenderer.renderAspect(guiGraphics, aspectStack, -AspectRenderer.PIXEL_RESOLUTION / 2, -AspectRenderer.PIXEL_RESOLUTION / 2);

      poseStack.popPose();
      i++;
    }

    guiGraphics.flush();
    poseStack.popPose();
  }

  public static void renderAspect(T7GuiGraphics guiGraphics, AspectStack aspectStack, int pX, int pY) {
    AspectRenderer.drawAspectIcon(guiGraphics, aspectStack.aspect(), pX, pY);
    AspectRenderer.drawCount(guiGraphics, String.valueOf(aspectStack.amount()), pX, pY);
  }

  public static void drawAspectIcon(T7GuiGraphics guiGraphics, Aspect aspect, int pX, int pY) {
    var sprite = AspectAtlas.sprite(Thavma.INSTANCE.rl(aspect.getId()));

    var color = aspect.getColor();
    RenderSystem.disableDepthTest();
    guiGraphics.blit(
        pX,
        pY,
        0,
        AspectRenderer.PIXEL_RESOLUTION,
        AspectRenderer.PIXEL_RESOLUTION,
        sprite,
        color
    );
    RenderSystem.enableDepthTest();
  }

  public static Vector2f[] calculateOffsets(int numAspects) {
    Vector2f[] offsets = new Vector2f[numAspects];

    int rows = (int) Math.ceil(1f * numAspects / AspectRenderer.ROW_SIZE);
    for (int i = 0; i < rows; i++) {
      int cols = Math.min(numAspects - (i * AspectRenderer.ROW_SIZE), AspectRenderer.ROW_SIZE);
      for (int j = 0; j < cols; j++) {
        float xOffset = (cols - 1) / -2f + j;
        offsets[i * AspectRenderer.ROW_SIZE + j] = new Vector2f(xOffset, i);
      }
    }

    return offsets;
  }

  public static void drawCount(T7GuiGraphics guiGraphics, String text, int pX, int pY) {
    var poseStack = guiGraphics.pose();
    poseStack.pushPose();
    poseStack.translate(pX + AspectRenderer.PIXEL_RESOLUTION, pY + AspectRenderer.PIXEL_RESOLUTION, 0.0001f); // start bottom right, like item count. slightly increase Z to avoid z fighting
    poseStack.scale(0.5F, 0.5F, 0.5F);
    Font font = Minecraft.getInstance().font;
    poseStack.translate(-font.width(text), -font.lineHeight, 0);
    GuiGraphicsExtensionsKt.drawOutlinedSeethroughString(guiGraphics, font, text, 0xFFFFFF, 0);
    poseStack.popPose();
  }

  public static int getPixelResolution() {
    return AspectRenderer.PIXEL_RESOLUTION;
  }
}

File: src/main/java/me/alegian/thavma/impl/client/renderer/AspectRenderer.kt
package me.alegian.thavma.impl.client.renderer

import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.math.Axis
import me.alegian.thavma.impl.client.texture.atlas.AspectAtlas
import me.alegian.thavma.impl.client.util.*
import me.alegian.thavma.impl.common.aspect.Aspect
import me.alegian.thavma.impl.common.aspect.AspectMap
import me.alegian.thavma.impl.common.aspect.AspectStack
import me.alegian.thavma.impl.common.util.vec2
import me.alegian.thavma.impl.rl
import net.minecraft.client.Camera
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.core.BlockPos
import net.minecraft.world.phys.Vec2
import net.minecraft.world.phys.Vec3
import me.alegian.thavma.impl.common.util.plus
import me.alegian.thavma.impl.common.util.unaryMinus
import kotlin.math.ceil
import kotlin.math.min

private const val ROW_SIZE: Int = 5 // max aspects per row
private const val QUAD_SIZE: Float = .3f // aspect width in-world is .3 blocks

/**
 * Renders Aspects, either in-world or in GUIs.
 * The draw call always has size 16x16, use pose manipulation to change size
 */
object AspectRenderer {
  fun renderAfterWeather(aspects: AspectMap, poseStack: PoseStack, camera: Camera, blockPos: BlockPos) {
    if (aspects.isEmpty) return

    val guiGraphics = GuiGraphics(Minecraft.getInstance(), poseStack, Minecraft.getInstance().renderBuffers().bufferSource())

    guiGraphics.usePose {
      translate(-camera.position + blockPos.center + Vec3(0.0, 0.75, 0.0))
      val angle = calculatePlayerAngle(blockPos.center)
      mulPose(Axis.YP.rotation(angle))

      scaleXY(QUAD_SIZE) // this puts us in "aspect space"
      scaleXY(1 / 16f) // this puts us in "pixel space"
      scale(1f, -1f, 1f) // world Y to GUI Y
      renderAspectsWrapped(aspects, guiGraphics)
    }
  }

  fun renderAspectsWrapped(aspects: AspectMap, guiGraphics: GuiGraphics) {
    // these offsets account for wrapping to new lines, and centering the aspects
    val offsets = calculateOffsets(aspects.size)

    for ((i, aspectStack) in aspects.withIndex())
      guiGraphics.usePose {
        translateXY(offsets[i])
        translateXY(-8, 0)
        renderAspect(guiGraphics, aspectStack)
      }
  }

  fun renderAspect(guiGraphics: GuiGraphics, aspectStack: AspectStack) {
    drawAspectIcon(guiGraphics, aspectStack.aspect)
    drawCount(guiGraphics, aspectStack.amount.toString())
  }

  fun drawAspectIcon(guiGraphics: GuiGraphics, aspect: Aspect) {
    val sprite = AspectAtlas.getSprite(rl(aspect.id))

    RenderSystem.disableDepthTest()
    guiGraphics.blit(
      0,
      0,
      0,
      16,
      16,
      sprite,
      aspect.color
    )
    RenderSystem.enableDepthTest()
  }

  private fun calculateOffsets(numAspects: Int): List<Vec2> {
    val offsets = MutableList(numAspects) { Vec2.ZERO }

    val rows = ceil((1f * numAspects / ROW_SIZE)).toInt()
    for (i in 0..<rows) {
      val cols = min((numAspects - (i * ROW_SIZE)), ROW_SIZE)
      for (j in 0..<cols) {
        val xOffset = (cols - 1) / -2f + j
        offsets[i * ROW_SIZE + j] = vec2(16 * xOffset, -16 * i - 8)
      }
    }

    return offsets
  }

  private fun drawCount(guiGraphics: GuiGraphics, text: String) {
    guiGraphics.usePose {
      translateXY(16, 16) // bottom right, like item count
      scale(0.5f)
      val font = Minecraft.getInstance().font
      translateXY(-font.width(text), -font.lineHeight)
      guiGraphics.drawOutlinedSeethroughString(font, text, 0xFFFFFF, 0)
    }
  }
}
